lib/cartopy/mpl/scale_bar.py

```diff
@@ -0,0 +1,411 @@
+
+
+import matplotlib
+matplotlib.use('Agg')
+
+# third party imports
+import matplotlib.font_manager as fm
+import matplotlib.patheffects as path_effects
+import matplotlib.pyplot as plt
+import numpy as np
+import pyproj
+import cartopy.crs as ccrs
+from cartopy.mpl.geoaxes import GeoAxes
+import cartopy
+
+import numpy as np
+
+
+# CONSTANTS
+BAR_HEIGHT_FONT = 0.75    # bar height in fontsize units
+MAX_BAR_HEIGHT = 0.5      # maximum bar height in inches
+POINTS_TO_INCHES = 1 / 72  # inches in one point
+TICK_FRACTION = 0.3       # how much of a fontsize should tick height be?
+GAP_FRACTION = 0.3        # how much of a fontsize should tick gap be?
+BAR_WIDTH_FRACTION = 0.25  # how wide should the default bar be
+DD2METERS = 111191.0
+METERS2MILES = 3.28 / 5280
+METERS2KM = 1 / 1000
+
+# the default set of lengths (keys) and number of divisions (values)
+LENGTHS = {4: 4,
+           10: 2,
+           12: 4,
+           15: 3,
+           20: 4,
+           30: 3,
+           40: 4,
+           75: 3,
+           100: 4,
+           120: 4,
+           150: 3,
+           200: 4,
+           240: 6,
+           300: 6,
+           400: 4,
+           600: 6,
+           750: 3,
+           1000: 4,
+           2000: 4,
+           4000: 4,
+           6000: 6,
+           10000: 4,
+           20000: 4,
+           40000: 4,
+           60000: 6,
+           75000: 3,
+           100000: 4,
+           200000: 4, }
+
+# Proj4 supported units and their conversion factor to meters
+UNITS = {'km': 1000.,
+         'm': 1.,
+         'dm': 1 / 10,
+         'cm': 1 / 100,
+         'mm': 1 / 1000,
+         'kmi': 1852.0,
+         'in': 0.0254,
+         'ft': 0.3048,
+         'yd': 0.9144,
+         'mi': 1609.344,
+         'fath': 1.8288,
+         'ch': 20.1168,
+         'link': 0.201168,
+         'us-in': 1. / 39.37,
+         'us-ft': 0.304800609601219,
+         'us-yd': 0.914401828803658,
+         'us-ch': 20.11684023368047,
+         'us-mi': 1609.347218694437,
+         'ind-yd': 0.91439523,
+         'ind-ft': 0.30479841,
+         'ind-ch': 20.11669506}
+
+
+class ScaleBar(object):
+    def __init__(self, ax, pos='ll', padx=0.1, pady=0.1,
+                 font_size=None, units='km',
+                 zorder=20, bar_length=None, divisions=None):
+        """Draw a scale bar on an axes, choosing the best defaults possible.
+
+    Default behaviors:
+     - The height of the bar will be relative to the font size.
+     - The length of the tick on the bottom of the bar will be 0.5 * font_size
+     - The gap between the tick and the top of the tick label will be
+       0.5 * font_size
+     - The width of the bar will be determined choosing from a list of lengths
+       the length that is closest to 0.25 * map width.
+     - The number of divisions will be associated with the chosen length.
+
+    Args:
+        ax: Cartopy GeoAxes object.
+        pos: String, one of ('ll','lr','ul','ur') (lower left, lower right,
```

Comment:
I've never been particularly fond of these kinds of positioning keywords. I wonder if we can improve on this by allowing users to create their own axes (with a helper function) to draw to.

Reply:
Perhaps this would be better implemented as a bounding box in Axes coordinates... (where users specify the left, right, top, bottom kwargs as necessary (and may specify any, all or none).